upload/library/Sidane/Threadmarks/Model/Threadmarks.php

```diff
@@ -3,59 +3,107 @@
 class Sidane_Threadmarks_Model_Threadmarks extends XenForo_Model
 {
 
-  public function createThreadMark($thread, $post, $label) {
+  public function SetThreadMark($thread_id, $post_id, $label) {
     $db = $this->_getDb();
-
+    
     XenForo_Db::beginTransaction($db);
-
-    $db->query('
-      INSERT IGNORE INTO threadmarks
+    
+    $stmt =  $db->query('
+      INSERT INTO threadmarks
         (thread_id, post_id, label)
       VALUES
         (?, ?, ?)
-    ', array($thread['thread_id'], $post['post_id'], $label));
-
-    $db->query('
-      UPDATE xf_thread
-      SET has_threadmarks = 1
-      WHERE thread_id = ?
-    ', $thread['thread_id']);
-
+      ON DUPLICATE KEY UPDATE  
+        label = values(label)
+    ', array($thread_id, $post_id, $label));
+    $rowsAffected = $stmt->rowCount();
+    
+    // http://dev.mysql.com/doc/refman/5.0/en/insert-on-duplicate.html
+    // 1 - new row, 2 - update
+    if ($rowsAffected == 1)
+    {
+        $db->query('
+          UPDATE xf_thread
+          SET threadmark_count = threadmark_count + 1
+          WHERE thread_id = ?
+        ', $thread_id);
+    }
+    
     XenForo_Db::commit($db);
 
     return true;
-  }
+  } 
 
-  public function deleteThreadMark($threadmark) {
-    $db = $this->_getDb();
+    public function deleteThreadMark($threadmark, $decrementCount = false) 
+    {
+        $db = $this->_getDb();
 
-    XenForo_Db::beginTransaction($db);
+        XenForo_Db::beginTransaction($db);
 
-    $db->query('
-      DELETE FROM threadmarks WHERE threadmark_id = ?
-    ', $threadmark['threadmark_id']);
+        $db->query('
+            DELETE FROM threadmarks WHERE threadmark_id = ?
+        ', $threadmark['threadmark_id']);
 
-    $threadmarks = $this->getByThreadId($threadmark['thread_id']);
+        if ($decrementCount)
+        {
+            $this->modifyThreadMarkCount($threadmark['thread_id'], -1);
+        }
+        
+        XenForo_Db::commit($db);
 
-    if (count($threadmarks) == 0) {
-      $db->query('
-        UPDATE xf_thread
-        SET has_threadmarks = 0
-        WHERE thread_id = ?
-      ', $threadmark['thread_id']);
+        return true;
     }
 
-    XenForo_Db::commit($db);
-
-    return true;
-  }
-
+    public function modifyThreadMarkCount($thread_id, $increment)
+    {
+        $db = $this->_getDb();
+        $db->query('
+            UPDATE xf_thread
+            SET threadmark_count = threadmark_count + ?
+            WHERE thread_id = ? and threadmark_count + ? > 0 
+        ', array($increment, $thread_id, $increment));
+    }
+  
+    public function rebuildThreadMarkCache($thread_id)
+    {
+        $db = $this->_getDb();
+        
+        // remove orphaned threadmarks
+        $db->query('
+            DELETE `threadmarks`
+            FROM `threadmarks`
+            LEFT JOIN xf_post AS post on post.thread_id = threadmarks.thread_id and post.post_id = threadmarks.post_id
+            where `threadmarks`.thread_id = ? 
+        ', $thread_id);
```

Comment:
This query deleted every existing threadmark record when it ran during the add-on upgrade.